ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -93,2565 +92,1438 @@
-	int n, s, r, d;
+	mbfl_string_init_set(result, string->encoding);
+	mbfl_memory_device device;
+	mbfl_memory_device_init(&device, string->len, 0);
+	mbfl_convert_filter *decoder = mbfl_convert_filter_new(&mbfl_encoding_wchar, string->encoding,
+		mbfl_memory_device_output, 0, &device);
 
-	if (pc->status) {
-		switch (pc->status) {
-		case 1: /* '&' */
-			(*pc->decoder->filter_function)(0x26, pc->decoder);		/* '&' */
-			break;
-		case 2: /* '#' */
-			(*pc->decoder->filter_function)(0x26, pc->decoder);		/* '&' */
-			(*pc->decoder->filter_function)(0x23, pc->decoder);		/* '#' */
-			break;
-		case 3: /* '0'-'9' */
-			(*pc->decoder->filter_function)(0x26, pc->decoder);		/* '&' */
-			(*pc->decoder->filter_function)(0x23, pc->decoder);		/* '#' */
-
-			s = pc->cache;
-			r = 1;
-			n = pc->digit;
-			while (n > 1) {
-				r *= 10;
-				n--;
-			}
-			while (r > 0) {
-				d = s/r;
-				s %= r;
-				r /= 10;
-				(*pc->decoder->filter_function)(mbfl_hexchar_table[d], pc->decoder);
-			}
+	size_t len;
+	unsigned int *wc_buffer = convert_string_to_wchar(string, &len), *p = wc_buffer, *e = wc_buffer + len;
 
-			break;
-		case 4: /* 'x' */
-			(*pc->decoder->filter_function)(0x26, pc->decoder);		/* '&' */
-			(*pc->decoder->filter_function)(0x23, pc->decoder);		/* '#' */
-			(*pc->decoder->filter_function)(0x78, pc->decoder);		/* 'x' */
-			break;
-		case 5: /* '0'-'9','a'-'f' */
-			(*pc->decoder->filter_function)(0x26, pc->decoder);		/* '&' */
-			(*pc->decoder->filter_function)(0x23, pc->decoder);		/* '#' */
-			(*pc->decoder->filter_function)(0x78, pc->decoder);		/* 'x' */
-
-			s = pc->cache;
-			r = 1;
-			n = pc->digit;
-			while (n > 0) {
-				r *= 16;
-				n--;
-			}
-			s %= r;
-			r /= 16;
-			while (r > 0) {
-				d = s/r;
-				s %= r;
-				r /= 16;
-				(*pc->decoder->filter_function)(mbfl_hexchar_table[d], pc->decoder);
-			}
-			break;
-		default:
-			break;
-		}
-	}
+continue_decoding_html_entity:
+	while (p < e) {
+		unsigned int c = *p++;
 
-	pc->status = 0;
-	pc->cache = 0;
-	pc->digit = 0;
+		if (c == '&' && p < e) {
+			/* OK, we have something which _might_ be an HTML numeric entity. Let's see what comes next... */
+			unsigned int *entity = p - 1; /* So we can go back and output these chars if it's not a valid numeric entity */
+			c = *p++;
 
-	return 0;
-}
+			if (c == '#' && p < e) {
+				c = *p++;
 
+				if (c == 'x' && p < e) {
+					/* Start decoding hexadecimal number */
+					int hex = 0, digits = 0;
+					c = *p++;
 
-mbfl_string *
-mbfl_html_numeric_entity(
-    mbfl_string *string,
-    mbfl_string *result,
-    int *convmap,
-    int mapsize,
-    int type)
-{
-	struct collector_htmlnumericentity_data pc;
-	mbfl_memory_device device;
-	mbfl_convert_filter *encoder;
-	size_t n;
-	unsigned char *p;
+					/* If very next character is not a hex digit, then what we have found
+					 * is not actually a hexadecimal numeric entity */
+					if ((c < '0' || c > '9') && (c < 'A' || c > 'F') && (c < 'a' || c > 'f')) {
```

Comment:
I'm wondering if it's possible to avoid this up-front check (which is kinda awkward due to goto into loop below). I.e. directly process first char as hex and possibly check digits == 0?

Reply:
Good idea. Implemented that. (At the same time, I realized that this function won't process an HTML numeric entity if it comes *immediately* after another one without a terminating `;`. I think I should probably fix that.)